julearn/pipeline/merger.py

```diff
@@ -0,0 +1,116 @@
+
+    Parameters
+    ----------
+    pipelines : List[EstimatorLike]
+        List of estimators that will be merged.
+    search_params : Dict
+        Dictionary with the search parameters.
+
+    Returns
+    -------
+    merged : BaseSearchCV
+        The merged pipeline as a searcher.
+    """
+
+    # Check that we only merge pipelines and searchers. And if there is a
+    # searcher, they are all of the same kind and match the search params.
+
+    search_params = prepare_search_params(search_params)
+
+    for p in pipelines:
+        if not isinstance(p, (Pipeline, GridSearchCV, RandomizedSearchCV)):
+            raise_error(
+                "Only pipelines and searchers are supported. "
+                f"Found {type(p)} instead.")
+        if isinstance(p, GridSearchCV):
+            if search_params["kind"] != "grid":
+                raise_error(
+                    "At least one of the pipelines to merge is a "
+                    "GridSearchCV, but the search params do not specify a "
+                    "grid search. These pipelines cannot be merged."
+                )
+        elif isinstance(p, RandomizedSearchCV):
+            if search_params["kind"] != "random":
+                raise_error(
+                    "At least one of the pipelines to merge is a "
+                    "RandomizedSearchCV, but the search params do not specify "
+                    "a random search. These pipelines cannot be merged."
+                )
+
+    # Check that all estimators have the same named steps in their pipelines.
+    reference_pipeline = pipelines[0]
+    if isinstance(reference_pipeline, (GridSearchCV, RandomizedSearchCV)):
+        reference_pipeline = reference_pipeline.estimator
+
+    step_names = reference_pipeline.named_steps.keys()
+
+    for p in pipelines:
+        if isinstance(p, (GridSearchCV, RandomizedSearchCV)):
+            p = p.estimator
+            if not isinstance(p, Pipeline):
+                raise_error("All searchers must use a pipeline.")
+        if step_names != p.named_steps.keys():
+            raise_error("All pipelines must have the same named steps.")
+
+    # The idea behind the merge is to create a list of parameter
+    # grids/distributions from a list of pipeline and searchers, to then
+    # wrap them into a single searcher. Since all the searchers have the same
+    # steps, this is possible. We just need to concatenate the
+    # grids/distributions from all searchers. If one of the pipelines is not
+    # a searcher, then this means that it has no hyperparameters to tune, but
+    # the pipeline is one of the hyperparameter options.
+
+    different_steps = []
+    for t_name in step_names:
+        # Get the transformer/model of the first element
+        t = reference_pipeline.named_steps[t_name]
+
+        # Check that all searchers have the same transformer/model.
+        # TODO: Fix this comparison, as it always returns False.
```

Comment:
probably something we can cover with tests and then find out whats wrong :+1:

Reply:
Indeed the problem is `!=` with scikit-learn models. I think they are testing for the reference or something like that, and not the actual parameters.